File: sdk/go/oauth.go
package qeetid

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type TokenExchangeInput struct {
	ClientID      string
	ClientSecret  string
	SubjectToken  string
	Scope         string // optional downscoped permissions
	ActorToken    string // optional — for RFC 8693 delegation
	ActorTokenType string
}

type TokenExchangeResult struct {
	AccessToken      string `json:"access_token"`
	TokenType        string `json:"token_type"`
	ExpiresIn        int    `json:"expires_in"`
	Scope            string `json:"scope,omitempty"`
	IssuedTokenType  string `json:"issued_token_type,omitempty"`
}

type IntrospectResult struct {
	Active    bool     `json:"active"`
	Sub       string   `json:"sub,omitempty"`
	Scope     string   `json:"scope,omitempty"`
	Exp       int64    `json:"exp,omitempty"`
	Iat       int64    `json:"iat,omitempty"`
	TenantID  string   `json:"tenant_id,omitempty"`
	ActorType string   `json:"actor_type,omitempty"`
	AgentID   string   `json:"agent_id,omitempty"`
}

// OAuth provides RFC 8693 token exchange, RFC 7662 token introspection, and
// an MCP token guard. Unlike other resources it uses form-encoded requests
// with OIDC client credentials rather than the shared API-key transport.
type OAuth struct {
	baseURL string
	hc      *http.Client
}

func newOAuth(baseURL string, hc *http.Client) *OAuth {
	return &OAuth{baseURL: baseURL, hc: hc}
}

// TokenExchange implements RFC 8693 downscoping and delegation.
func (o *OAuth) TokenExchange(ctx context.Context, in TokenExchangeInput) (*TokenExchangeResult, error) {
	params := url.Values{
		"grant_type":           {"urn:ietf:params:oauth:grant-type:token-exchange"},
		"subject_token":        {in.SubjectToken},
		"subject_token_type":   {"urn:ietf:params:oauth:token-type:access_token"},
		"requested_token_type": {"urn:ietf:params:oauth:token-type:access_token"},
	}
	if in.Scope != "" {
		params.Set("scope", in.Scope)
	}
	if in.ActorToken != "" {
		params.Set("actor_token", in.ActorToken)
		typ := in.ActorTokenType
		if typ == "" {
			typ = "urn:ietf:params:oauth:token-type:access_token"
		}
		params.Set("actor_token_type", typ)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, o.baseURL+"/v1/oauth/token", strings.NewReader(params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("qeetid oauth: build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.SetBasicAuth(in.ClientID, in.ClientSecret)

	res, err := o.hc.Do(req)
	if err != nil {
		return nil, fmt.Errorf("qeetid oauth: token exchange: %w", err)
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 1<<20))
	var out TokenExchangeResult
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, fmt.Errorf("qeetid oauth: decode response: %w", err)
	}
	if res.StatusCode >= 300 {
		return nil, parseError(res.StatusCode, body, res.Header.Get("X-Request-Id"), 0)
	}
	return &out, nil
}

// Introspect implements RFC 7662 token introspection.
func (o *OAuth) Introspect(ctx context.Context, token string) (*IntrospectResult, error) {
	params := url.Values{"token": {token}}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, o.baseURL+"/v1/oauth/introspect", strings.NewReader(params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("qeetid oauth: build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	res, err := o.hc.Do(req)
	if err != nil {
		return nil, fmt.Errorf("qeetid oauth: introspect: %w", err)
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 1<<20))
	if res.StatusCode >= 300 {
		return nil, parseError(res.StatusCode, body, res.Header.Get("X-Request-Id"), 0)
	}
	var out IntrospectResult
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, fmt.Errorf("qeetid oauth: decode response: %w", err)
	}
	return &out, nil
}

// Verify is an MCP token guard: introspects the token and returns an error if
// it is inactive or does not carry requiredScope (empty string skips scope check).
func (o *OAuth) Verify(ctx context.Context, token, requiredScope string) (*IntrospectResult, error) {
	result, err := o.Introspect(ctx, token)
	if err != nil {
		return nil, err
	}
	if !result.Active {
		return nil, &Error{Status: 401, Code: "token_inactive", Message: "token is not active"}
	}
	if requiredScope != "" {
		scopes := strings.Fields(result.Scope)
		found := false
		for _, s := range scopes {
			if s == requiredScope {
				found = true
				break
			}
		}
		if !found {
			return nil, &Error{Status: 403, Code: "insufficient_scope", Message: "required scope: " + requiredScope}
		}
	}
	return result, nil
}

File: sdk/js/node/src/oauth.ts
import { QeetIDError } from "./errors.js";
import type { FetchLike } from "./client.js";

export interface TokenExchangeInput {
  /** OIDC client_id that has the token_exchange grant. */
  clientId: string;
  /** OIDC client_secret. */
  clientSecret: string;
  /** The access token to exchange. */
  subjectToken: string;
  /** Downscoped space-separated permissions. Omit to inherit all subject scopes. */
  scope?: string;
  /** Actor token for RFC 8693 delegation (`act` claim). */
  actorToken?: string;
  actorTokenType?: string;
}

export interface TokenExchangeResult {
  access_token: string;
  token_type: string;
  expires_in: number;
  scope?: string;
  issued_token_type?: string;
}

export interface IntrospectResult {
  active: boolean;
  sub?: string;
  scope?: string;
  aud?: string | string[];
  exp?: number;
  iat?: number;
  tenant_id?: string;
  actor_type?: string;
  agent_id?: string;
  act?: Record<string, unknown>;
}

/** OAuth + MCP helpers: token-exchange (RFC 8693), introspect (RFC 7662), and an MCP token guard. */
export class OAuth {
  constructor(private readonly baseUrl: string, private readonly fetchImpl: FetchLike) {}

  /** RFC 8693 token exchange — downscope or add an `act` delegation claim. */
  async tokenExchange(input: TokenExchangeInput): Promise<TokenExchangeResult> {
    const params = new URLSearchParams({
      grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
      subject_token: input.subjectToken,
      subject_token_type: "urn:ietf:params:oauth:token-type:access_token",
      requested_token_type: "urn:ietf:params:oauth:token-type:access_token",
    });
    if (input.scope) params.set("scope", input.scope);
    if (input.actorToken) params.set("actor_token", input.actorToken);
    if (input.actorTokenType) {
      params.set("actor_token_type", input.actorTokenType);
    }
    const credentials = btoa(`${input.clientId}:${input.clientSecret}`);
    const res = await this.fetchImpl(`${this.baseUrl}/v1/oauth/token`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Authorization: `Basic ${credentials}`,
        Accept: "application/json",
      },
      body: params.toString(),
    });
    const data = await res.json() as Record<string, unknown>;
    if (!res.ok) {
      throw new QeetIDError(
        res.status,
        (data["error"] as string) ?? "token_exchange_failed",
        (data["error_description"] as string) ?? "Token exchange failed",
      );
    }
    return data as unknown as TokenExchangeResult;
  }

  /** RFC 7662 token introspection — check active state, scopes, and MCP actor claims. */
  async introspect(token: string): Promise<IntrospectResult> {
    const res = await this.fetchImpl(`${this.baseUrl}/v1/oauth/introspect`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json",
      },
      body: new URLSearchParams({ token }).toString(),
    });
    if (!res.ok) {
      throw new QeetIDError(res.status, "introspect_failed", "Token introspection failed");
    }
    return res.json() as Promise<IntrospectResult>;
  }

  /**
   * MCP token guard — verify a token is active and optionally has a required scope.
   * Throws QeetIDError(401) if inactive, QeetIDError(403) if scope is missing.
   * Use this in your MCP tool handlers to enforce authentication.
   */
  async verify(token: string, requiredScope?: string): Promise<IntrospectResult> {
    const result = await this.introspect(token);
    if (!result.active) {
      throw new QeetIDError(401, "token_inactive", "Token is not active");
    }
    if (requiredScope) {
      const scopes = (result.scope ?? "").split(" ").filter(Boolean);
      if (!scopes.includes(requiredScope)) {
        throw new QeetIDError(
          403,
          "insufficient_scope",
          `Required scope: ${requiredScope}`,
        );
      }
    }
    return result;
  }
}
